Removing leftover halts

dupriezt opened the pull request:
Or adding the <haltOrBreakpointForTesting> to methods that are meant to contain halts in production. This way they won't appear in the breakpoint browser that got integrated into Pharo 7 with the new version of Calypso.

Commit:
Removed leftover halt in FilterTest>>setUp

Commit:
Added pragma <haltOrBreakpointForTesting> to these methods, to hide their halts from the breakpoint browser.

Commit:
Added <haltOrBreakpointForTesting> pragma to these methods to hide them from the breakpoint browser

Commit:
Added <haltOrBreakpointForTesting> pragma to this methods to hide them from the breakpoint browser

Commit:
Added <haltOrBreakpointForTesting> pragma to these methods to hide them from the breakpoint browser

Commit:
Rewrote these tests not to use halt to test for reachability

Commit:
Removed useless halts

Commit:
Added <haltOrBreakpointForTesting> pragma to this methods to hide them from the breakpoint browser

Commit:
Added <haltOrBreakpointForTesting> pragma to these methods to hide them from the breakpoint browser

tesonep commented:
@dupriezt there is an issue in fogbugz for this? It should be integrated in Pharo7? or Pharo8?

guillep commented on src/Kernel-Tests/ProtoObjectTest.class.st:
```diff
@@ -78,19 +106,38 @@ ProtoObjectTest >> testIfNotNil [
+	block := [ :o | reached := true ].
+	object ifNotNil: block ifNil: [ notReached := true ].
+	self assert: reached equals: true.
+	self assert: notReached equals: false.
```
Is really all this repeated code needed?

dupriezt replied:
It's not testing exactly the same thing each time, so... maybe?
I don't know :-)

guillep commented on src/Manifest-Resources-Tests/MFClassA.class.st:
```diff
@@ -10,5 +10,5 @@ Class {
 MFClassA >> method [
 	|foo|
-	self halt.
+	1+1.
```
Why not just leaving an empty method returning `self`?

dupriezt replied:
I don't know. Do you want me to commit an empty method instead?

guillep replied:
An empty method would indeed be better.
Moreover, putting a message saying why that method is empty even better.

guillep replied:
Imagine in a couple of months the person that will read that 1+1 method that does not return and think it's a bug.

dupriezt replied:
Done 1227d92

guillep commented on src/Manifest-Resources-Tests/MFClassB.class.st:
```diff
@@ -15,5 +15,5 @@ MFClassB >> method2 [
 MFClassB >> method3 [
 	|foo|
-	self halt.
+	1+1.
```
Why not just leaving an empty method returning `self`?

dupriezt replied:
I don't know. Do you want me to commit an empty method instead?

guillep replied:
An empty method would indeed be better.
Moreover, putting a message saying why that method is empty even better.

guillep replied:
Imagine in a couple of months the person that will read that 1+1 method that does not return and think it's a bug.

dupriezt replied:
Done 1227d92

jecisc commented:
Since it make the Breakpoint browser more usable I think it night be good for P7.

guillep commented:
Let's wait until all tests run. I don't know if this will have an effect on them or not.

dupriezt commented:
Les deux tests qui ne passent pas sont:
- unix-32 / Tests-unix-32 / testSizeAndSpan – Unix32.Glamour.Tests.Morphic.GLMTabulatorMorphicTest
- windows-32 / Tests-windows-32 / testAliasCompositions – Windows32.Tests.Traits.TraitCompositionTest

J'ai testé le premier sur mon linux, et il passe...
Que faut-il faire?

jecisc commented:
They are unrelated. I can relaunch it later.

tesonep commented:
This PR should be carefully addressed. It is changing the semantics of the tests.

The tests are not well written, but they are using the halts instead of asserts or failures. 

For example, the proto object test is testing that the blocks are not executed in if the receiver is not nil, and it is testing it with and without parameter in the block. 

A test should not be modified without understanding it. Unless we have tests for the tests.

I rather prefer a false positive in a tool than losing a test

dupriezt commented:
I'm not an expert, but it seems that the intention of these tests on ProtoObject was just to check that the correct branch of the ifNil (and flavours) was taken each time. Which is preserved by using booleans to track which branch was taken.
Do you think I should re-write these tests to use another kind of exceptions than halts and errors to check which branches are taken?

guillep commented:
I'd be ok to integrate this. But we should make an issue to enhance it in Pharo8 I think.

For example, notReached := true  can be just `self fail`

Commit:
Made these methods empty instead of having dummy "1+1" code

dupriezt commented:
I removed the dummy code from the methods Guille pointed at earlier in this discussion. These methods are now empty.

guillep reviewed (changes requested):
We need a fogbugz issue for this PR, and the branch has to be named correspondingly, otherwise the CI server cannot correctly send the corresponding issues. This has been said a week ago.

Also, the PR has broken a couple of tests, probably related to the fact of leaving unused temporary variables.

Commit:
Fix to make tests pass

dupriezt commented:
Thanks for the list of failing tests @guillep I looked into them:
- testCriticsOf fails because I removed the unused variables from MFClassA>>method, MFClassB>>method2 and MFClassB>>method3 and this test was checking that these methods were indeed triggering a "variable not read or written" critics rule. I fixed that in my last commit, and added a comment to these methods.
- testIsToDo fails because... I'm not really sure why. But adding a line of code to MFClassB>>method2 makes it pass so I did that and added a comment to the method.
- same thing for testToDoOf
- testNoUnusedTemporaryVariablesLeft fails because of the way it is written. It checks that there are not temporary variables unused in the system. To do that it builds the list of all methods with unused temporaries, remove from it a list of exceptions (method, method2 and method3 from above) from it and checks that the resulting list is empty. So when I removed the unused variables from these methods, the removeAll: fails because it tries to remove elements that are not in the list and that's what makes the test fail. It now passes with my last commit that put back the unused variables in method, method2 and method3, but I think there's maybe a design problem with the way this test is written.
